### PR TITLE
Fix error handling of Channel::Close at all call sites 

### PR DESCRIPTION
**Pull request by bocon13**
This PR fixes the error handling logic around `Channel::Close` calls. Previously the `APPEND_ERROR(status) << "msg"` pattern was used, but this did not account for the case where `status` might be OK, as `Close` returns a `bool`. We replace it with the `MAKE_ERROR(ERR_INTERNAL) << "msg"` followed by `APPEND_STATUS_IF_ERROR(status, error)` pattern, constructing a new error status. This pattern is well established throughout the code base (grep `= MAKE_ERROR`).

This is an extension to https://github.com/stratum/stratum/commit/177d91a691d43f1d31420514f5b435bf80f10ac8 which did not address the underlying issue.

    Co-authored-by: Maximilian Pudelko pudelkoM@users.noreply.github.com
    Co-authored-by: Yi Tseng yi@opennetworking.org

### CODE REVIEW

File: stratum/hal/lib/common/gnmi_publisher.cc [hunk 1]
```diff
@@ -311,9 +310,10 @@ ::util::Status GnmiPublisher::UnregisterEventWriter() {
                            switch_interface_->UnregisterEventNotifyWriter());
     APPEND_STATUS_IF_ERROR(status, parse_tree_.UnregisterEventNotifyWriter());
     // Close Channel.
-    if (!event_channel_ || !event_channel_->Close()) {
-      status = APPEND_ERROR(status)
-               << " Event Notify Channel is already closed.";
+    if (!event_channel_->Close()) {
+      ::util::Status new_status = MAKE_ERROR(ERR_INTERNAL)
```
Comment (bocon13) [inline stratum/hal/lib/common/gnmi_publisher.cc:314]:
```suggestion
      ::util::Status error = MAKE_ERROR(ERR_INTERNAL)
```

File: stratum/hal/lib/common/gnmi_publisher.cc [hunk 2]
```diff
@@ -311,9 +310,10 @@ ::util::Status GnmiPublisher::UnregisterEventWriter() {
...
+                                  << " Event Notify Channel is already closed.";
+      APPEND_STATUS_IF_ERROR(status, new_status);
     }
```
Comment (bocon13) [inline stratum/hal/lib/common/gnmi_publisher.cc:317]:
```suggestion
    if (!event_channel_->Close()) {
      ::util::Status error = MAKE_ERROR(ERR_INTERNAL)
                                  << " Event Notify Channel is already closed.";
      APPEND_STATUS_IF_ERROR(status, error);
    }
```

**Review by Yi-Tseng — APPROVED** [review]

File: stratum/hal/lib/p4/p4_table_mapper.cc [hunk 3]
```diff
@@ -712,21 +711,16 @@ ::util::Status P4TableMapper::ProcessMatchField(
...
   auto convert_iter = field_convert_by_table_.find(key);
   if (convert_iter == field_convert_by_table_.end()) {
-    ::util::Status field_error = MAKE_ERROR(ERR_OPER_NOT_SUPPORTED)
-                                 << "P4 TableEntry match field ID "
-                                 << PrintP4ObjectID(match_field.field_id())
-                                 << " is not recognized in table "
-                                 << table_p4_info.preamble().name();
-    APPEND_STATUS_IF_ERROR(status, field_error);
-    return status;  // No way to decode fields that don't go with the table.
+    RETURN_ERROR(ERR_OPER_NOT_SUPPORTED)
```
Comment (bocon13) [inline stratum/hal/lib/p4/p4_table_mapper.cc:720]:
```suggestion
    // No way to decode fields that don't go with the table.
    RETURN_ERROR(ERR_OPER_NOT_SUPPORTED)
```

**Comment by pudelkoM** [comment]
@bocon13 digging through the other `Channel::Close` call site code, I think you found a genuine bug. Nowhere was this error handled correctly.
https://github.com/stratum/stratum/commit/177d91a691d43f1d31420514f5b435bf80f10ac8 changed it from 
```c++
    APPEND_ERROR(status)
        << "Error when closing port status change event channel.";
```
to
```c++
    status = APPEND_ERROR(status)
             << "Error when closing port status change event channel.";
```

but did not address the underlying issue. Curious how all the `APPEND_ERROR` discards where in the context of `->Close()`, too.

**Review by pudelkoM — APPROVED** [review]

**Review by Yi-Tseng — DISMISSED** [review]

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 4]
```diff
@@ -852,17 +852,20 @@ ::util::Status BFChassisManager::UnregisterEventWriters() {
...
     xcvr_event_writer_id_ = kInvalidWriterId;
     if (!xcvr_event_channel_->Close()) {
-      status = APPEND_ERROR(status)
-               << "Error when closing transceiver event channel.";
+      ::util::Status error = MAKE_ERROR(ERR_INTERNAL)
+                             << "Error when closing transceiver event channel.";
+      APPEND_STATUS_IF_ERROR(status, error);
     }
   } else {
     return MAKE_ERROR(ERR_INTERNAL)
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:871]: Remove this else block, or make this an append

**Review by pudelkoM — APPROVED** [review]
Good catches in the last commits!

Fixed the last compiler errors, this looks good.

**Review by pudelkoM — APPROVED** [review]

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 5]
```diff
@@ -956,22 +965,25 @@ ::util::Status BFChassisManager::UnregisterEventWriters() {
...
-    status = APPEND_ERROR(status)
-             << "Error when closing port status change event channel.";
+  if (!port_status_change_event_channel_ ||
+      !port_status_change_event_channel_->Close()) {
+    ::util::Status error = MAKE_ERROR(ERR_INTERNAL)
+                           << "Error when closing port status change"
+                           << " event channel.";
+    APPEND_STATUS_IF_ERROR(status, error);
   }
+  port_status_change_event_channel_.reset();
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:975]:
```suggestion
  // TODO(bocon): need nullptr check
  port_status_change_event_channel_.reset();
```

**Pull request merged**